### PR TITLE
RSA SHA2 256/512 key upgrade support libssh2 #536

### PR DESCRIPTION
**Pull request by willco007**
Files: openssl.h, openssl.c, crypto.h, hostkey.c, HACKING-CRYPTO

Notes:
Added rsa-sha2-256 and rsa-sha2-512 as described in RFC 8332 to OpenSSL backend. To implement in other backends implementers need to define LIBSSH2_RSA_SHA2 and implement _libssh2_rsa_sha2_sign and _libssh2_rsa_sha2_verify.

Credit:
Will Cosgrove

### CODE REVIEW

**Comment by palmin** [comment]
@willco007 I'm not able to find `ext-info-c` mentioned anywhere which makes me wonder if the servers will ever send the `SSH_MSG_EXT_INFO` packet.

This is the change I have in `kex.c`:
https://github.com/palmin/libssh2/commit/d43a8d40fd295d8c5c0c97496726350af2dc1cf6#diff-2060d014fa38220550b67c00d442ce6b1b40cbf351d106913a4fb55f35b4212fR3029-R3040

**Comment by palmin** [comment]
@willco007 I have looked through the changed files and compiled https://github.com/willco007/libssh2/commit/9d1a3a5189ef2531978695aeee8a1322ab984cf3 where I can verify that `rsa-sha2-512` is used.

Looks good.

**Comment by palmin** [comment]
@willco007 I have met servers where the hostkey type passed to `hostkey_method_ssh_rsa_init()` is `rsa-sha2-256` or `rsa-sha2-512` instead of the expected `ssh-rsa` causing negotiation to fail.

Here are some adjustments taking this into account: https://github.com/palmin/libssh2/commit/3579f212bd2b9338979f4cc499f5132366a15239

**Comment by willco007** [comment]
@palmin good catch, will update PR.

**Comment by palmin** [comment]
Anything I can do to help merge this pull request?

Do we need additional people to review it?

**Comment by riggs** [comment]
Give that github is deprecating SHA-1 RSA today, 2022-1-11, and `ssh-keygen`'s default is SHA-2 RSA, I hope that this feature can be added to a patch release soon.

File: src/packet.c [hunk 1]
```diff
@@ -615,6 +615,75 @@ _libssh2_packet_add(LIBSSH2_SESSION * session, unsigned char *data,
...
+                    }
+
+                    if(name_len == 15 &&
+                        memcmp(name, "server-sig-algs", 15) == 0) {
+                        if(session->server_sign_algorithms) {
+                            LIBSSH2_FREE(session,
+                                         session->server_sign_algorithms);
+                        }
+
+                        session->server_sign_algorithms =
```
Comment (tsengjun) [inline src/packet.c:666]: Hi Will, @willco007
It seems the "server_sign_algorithms" was not freed at the session end.
Can you have a look?
Thanks!

**Comment by wastez** [comment]
I setted in the .h files the LIBSSH2_RSA_SHA2 to 1 and LIBSSH2_RSA to 0 but it isnt working

**Comment by willco007** [comment]
@wastez The `LIBSSH2_RSA_SHA2` and `LIBSSH2_RSA` flags are not something you normally change. These flags tell the library the crypto backend has implemented the feature. Unless you're implementing the RSA SHA2 backend you wouldn't change these values. Lastly, `LIBSSH2_RSA_SHA2` relies on  `LIBSSH2_RSA` being enabled.

**Comment by wastez** [comment]
If i enable both of them compilation runs into a error.
/usr/bin/ld: ../src/.libs/libssh2.so: undefined reference to `_libssh2_rsa_sha2_verify'
/usr/bin/ld: ../src/.libs/libssh2.so: undefined reference to `_libssh2_rsa_sha2_sign'
collect2: error: ld returned 1 exit status

What i´ve to do to use rsa-sha2-256 and rsa-sha2-512.

**Pull request merged**